### PR TITLE
Wizard recipe: ranger-v0.13.1

### PR DESCRIPTION
**Pull request by tecosaur**
This pull request contains a new build recipe I built using the BinaryBuilder.jl wizard:

* Package name: ranger
* Version: v0.13.1

@staticfloat please review and merge.

### CODE REVIEW

**Comment by tecosaur** [comment]
BinaryBuilder didn't pick up on any licence file, because there isn't any, but the C++ files start with comments that declare them to be under the MIT licence.

https://github.com/imbs-hl/ranger/blob/9aac567f21ef45f84b05ed53d2909e243e5ae7c1/cpp_version/src/main.cpp#L8
> the C++ core of ranger is distributed under MIT license

Please advise whether any action is needed.

File: R/ranger/bundled/patches/make_lib_and_install.patch [hunk 1]
```diff
@@ -0,0 +1,44 @@
+diff --git before/ranger-0.13.1/cpp_version/CMakeLists.txt after/ranger-0.13.1/cpp_version/CMakeLists.txt
+index 397bb53..34ee308 100644
+--- before/ranger-0.13.1/cpp_version/CMakeLists.txt
++++ after/ranger-0.13.1/cpp_version/CMakeLists.txt
```
Comment (giordano) [inline R/ranger/bundled/patches/make_lib_and_install.patch:4]:
```suggestion
--- before/cpp_version/CMakeLists.txt
+++ after/cpp_version/CMakeLists.txt
```

File: R/ranger/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    DirectorySource("./bundled")
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir
+for f in ${WORKSPACE}/srcdir/patches/*.patch; do
+    atomic_patch -p1 ${f}
+done
+cd ranger-0.13.1/cpp_version/
```
Comment (giordano) [inline R/ranger/build_tarballs.jl:20]:
```suggestion
cd $WORKSPACE/srcdir/ranger*/cpp_version/
for f in ${WORKSPACE}/srcdir/patches/*.patch; do
    atomic_patch -p2 ${f}
done
```

File: R/ranger/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+cmake -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIB=1
+make -j$nproc
+make install
+exit
```
Comment (giordano) [inline R/ranger/build_tarballs.jl:24]: > Please advise whether any action is needed.

We have the text of the MIT license in `/usr/share/licenses/`
```suggestion
install_license /usr/share/licenses/MIT
```

File: R/ranger/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    Platform("powerpc64le", "linux"; libc = "glibc"),
+    Platform("i686", "linux"; libc = "musl"),
+    Platform("x86_64", "linux"; libc = "musl"),
+    Platform("aarch64", "linux"; libc = "musl"),
+    Platform("armv6l", "linux"; call_abi = "eabihf", libc = "musl"),
+    Platform("armv7l", "linux"; call_abi = "eabihf", libc = "musl"),
+    Platform("x86_64", "macos"; ),
+    Platform("aarch64", "macos"; )
+]
+
```
Comment (giordano) [inline R/ranger/build_tarballs.jl:44]:
```
┌ Warning: /agent/_work/1/s/R/ranger/build/x86_64-linux-gnu/q8Sqg58m/x86_64-linux-gnu-libgfortran3-cxx03/destdir/lib/libranger.so contains std::string values!  This causes incompatibilities across the GCC 4/5 version boundary.  To remedy this, you must build a tarball for both GCC 4 and GCC 5.  To do this, immediately after your `platforms` definition in your `build_tarballs.jl` file, add the line:
│ 
│     platforms = expand_cxxstring_abis(platforms)
└ @ BinaryBuilder.Auditor /depot/packages/BinaryBuilder/9VQVb/src/auditor/compiler_abi.jl:247
```
```suggestion
platforms = expand_cxxstring_abis(platforms)
```

File: R/ranger/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    Platform("armv7l", "linux"; call_abi = "eabihf", libc = "glibc"),
+    Platform("powerpc64le", "linux"; libc = "glibc"),
+    Platform("i686", "linux"; libc = "musl"),
+    Platform("x86_64", "linux"; libc = "musl"),
+    Platform("aarch64", "linux"; libc = "musl"),
+    Platform("armv6l", "linux"; call_abi = "eabihf", libc = "musl"),
+    Platform("armv7l", "linux"; call_abi = "eabihf", libc = "musl"),
+    Platform("x86_64", "macos"; ),
+    Platform("aarch64", "macos"; )
+]
```
Comment (giordano) [inline R/ranger/build_tarballs.jl:43]:
```suggestion
platforms = supported_platforms(;exclude=Sys.iswindows)
```

Comment (giordano) [reply R/ranger/build_tarballs.jl:42]: I don't know what was the problem, let's see?

Comment (tecosaur) [reply R/ranger/build_tarballs.jl:42]: I'm afraid I don't have a record. Would the thing to do be commit your suggestion and look for CI failures?

Comment (giordano) [reply R/ranger/build_tarballs.jl:42]: Yes

**Comment by giordano** [comment]
```
[16:34:42] /opt/bin/x86_64-unknown-freebsd12.2-libgfortran3-cxx11/x86_64-unknown-freebsd12.2-clang++ --sysroot=/opt/x86_64-unknown-freebsd12.2/x86_64-unknown-freebsd12.2/sys-root/  -Wall -std=c++11 -O3 -DNDEBUG CMakeFiles/ranger-bin.dir/src/main.o -o ranger  -Wl,-rpath,/workspace/srcdir/ranger-0.13.1/cpp_version libranger.so 
[16:34:42] libranger.so: undefined reference to `pthread_create'
```
It's missing `-pthread`, that looks like an upstream bug to me

**Comment by tecosaur** [comment]
For now would it be best to just leave out the freebsd build?

**Comment by giordano** [comment]
https://github.com/imbs-hl/ranger/blob/9aac567f21ef45f84b05ed53d2909e243e5ae7c1/cpp_version/CMakeLists.txt#L20-L23
```
## ======================================================================================##
## In Clang phtread flag only for compiler, not for linker. For
## windows use static linking
## ======================================================================================##
```

**Comment by tecosaur** [comment]
Right. I must confess that I'm not sure what to make of that — I don't do much with C/C++/make.

**Comment by giordano** [comment]
Even though they should do something like https://stackoverflow.com/a/29871891/2442087
```cmake

File: R/ranger/bundled/patches/make_lib_and_install.patch [hunk 6]
```diff
@@ -0,0 +1,44 @@
+diff --git before/ranger-0.13.1/cpp_version/CMakeLists.txt after/ranger-0.13.1/cpp_version/CMakeLists.txt
```
Comment (giordano) [inline R/ranger/bundled/patches/make_lib_and_install.patch:1]: It'd be nice to upstream your patch, also to reduce our maintenance burden.  Building the shared library can be useful also for others.

**Comment by tecosaur** [comment]
Thanks!

**Pull request merged**